src/test-util.ts

```diff
@@ -32,3 +34,195 @@ export function b(
+      module = await import(window.origin + '/' + path);
+    } catch (e) {
+      console.log('Error importing ' + path + ':', e);
+    }
+  }
+
+  const handle = async (event: MessageEvent) => {
+    if (event.source === null || event.source != window.opener) {
+      return;
+    }
+    const {id, expr} = event.data;
+    const response: Response = {tabId, id};
+    if (expr) {
+      try {
+        response.result = await AsyncFunction(
+          'module',
+          '"use strict"; Object.assign(self, module); {' + expr + '}',
+        )(module);
+      } catch (e) {
+        response.error = e;
+      }
+    }
+    event.source.postMessage(response);
+  };
+
+  window.addEventListener('message', handle, false);
+  void handle(event);
+}
+
+export type Tab = {
+  run: (expr: string) => Promise<unknown>;
+  close: () => void;
+};
+
+// Registered callbacks for listener(), by tabId and call id.
+const callbacks: Record<
+  string,
+  Record<
+    string,
+    {
+      resolve: (res: unknown) => void;
+      reject?: (err: unknown) => void;
+    }
+  >
+> = {};
+
+function listener(event: MessageEvent<Response>): void {
+  const {tabId, id, result, error} = event.data;
+  const tabCallbacks = callbacks[tabId];
+  if (tabCallbacks === undefined) {
+    return;
+  }
+  if (id == 'log') {
+    const levels: Record<string, typeof console.log> = {
+      error: console.error,
+      warn: console.warn,
+      info: console.info,
+      debug: console.debug,
+    };
+    const {level, message} = result as LogResponse;
+    (levels[level] ?? console.log)('Page:', message);
+  } else {
+    const {resolve, reject} = tabCallbacks[id];
+    if (error !== undefined && reject !== undefined) {
+      reject(error);
+    } else {
+      resolve(result);
+    }
+  }
+}
+
+export async function newTab(filename?: string): Promise<Tab> {
+  const tabId = uuid();
+  const {promise, resolve} = resolver<unknown>();
+
+  if (!Object.keys(callbacks).length) {
+    window.addEventListener('message', listener, false);
+  }
+  callbacks[tabId] = {init: {resolve}};
+  const tabCallbacks = callbacks[tabId];
+
+  // Open a new tab to a URL that doesn't run anything on load.
+  const tab = window.open(
+    document.location.origin + '/src/test-util.ts',
+    '_blank',
+  );
+  if (tab == null) {
+    throw new Error('Failed to open window');
+  }
+
+  // Add a script element to import this file and call initializeNewTab().
+  const script = document.createElement('script');
+  script.type = 'text/javascript';
+  script.text = `
+    const initialize = async (event) => {
+      let testutil = await import(window.origin + '/src/test-util.ts');
+      testutil.initializeNewTab(event);
+      window.removeEventListener('message', initialize, false);
+    };
+    window.addEventListener("message", initialize, false);
+  `;
+  tab.document.head.appendChild(script);
+
+  // Start initialization process, waiting for page to become live.
+  const interval = setInterval(() => {
```

Comment:
The `script` element has an onload event. Not sure if that helps here though?

Reply:
I played with this a bunch, and was unable to prove to myself that it ever fired. I'm not confident it wouldn't work, but am giving up for now.